### PR TITLE
Feature/sweep outcome metrics

### PR DESCRIPTION
**Pull request by tboam**
**Goals (and why)**: To gather metrics on what happens with each iteration of sweep.  Ultimately I want to be able to create alarms if sweep is disabled for a long time, or the lock is always out or the DB isn't online.

**Implementation Description (bullets)**:  Refactored `BackgroundSweeperImpl.java` to return a clear outcome (enum is at the bottom of the class) from each of its decision points.  This outcome is then used to change the `batchSizeMultiplier`, sleep duration and update metrics.

**Concerns (what feedback would you like?)**: `SweepOutcomeMetrics` is really simple for now as it just records the last outcome, when metrics are retrieved it returns 1 for the last run and 0 for everything else.  Since multiple sweep runs may occur between metric snapshots (currently every 2 minutes but I don't think this is guaranteed) the metric is really just a sample.  I think this allows me to create the alarms I want but is this over simplifying the solution?

I could use a `Meter` but this will give a weighted mean per second for each metric which doesn't seem to fit the use case here.  Another option would be to implement my own `Counter` which can reset between calls to `getCount` so you get exactly how many times each `SweepOutcome` occurred since the last time metrics were retrieved at the cost some additional complexity.

**Where should we start reviewing?**: `BackgroundSweeperImpl.java`

**Priority (whenever / two weeks / yesterday)**: next couple of days, would ideally like to get this in a release next week so products start picking it up and we can create monitors.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2531)

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 1]
```diff
@@ -172,15 +195,15 @@ private long getBackoffTimeWhenSweepHasNotRun() {
...
-    boolean runOnce() {
+    SweepOutcome runOnce() {
         Optional<TableToSweep> tableToSweep = getTableToSweep();
         if (!tableToSweep.isPresent()) {
             // Don't change this log statement. It's parsed by test automation code.
             log.debug("Skipping sweep because no table has enough new writes to be worth sweeping at the moment.");
-            return false;
+            return SweepOutcome.NOTHING_TO_SWEEP;
         } else {
             specificTableSweeper.runOnceAndSaveResults(tableToSweep.get());
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: This can throw a runtimeException, we should catch it and set OutCome as `ERROR`.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 2]
```diff
@@ -208,26 +231,25 @@ boolean runOnce() {
...
+
             if (!progress.isPresent() || tables.contains(progress.get().tableRef())) {
-                return false;
+                log.warn("The background sweep job failed unexpectedly; will retry with a lower batch size...",
+                        originalException);
+                return SweepOutcome.RETRYING_WITH_SMALLER_BATCH;
             } else {
                 specificTableSweeper.getSweepProgressStore().clearProgress();
-                return true;
+                log.info("The table being swept by the background sweeper was dropped, moving on...");
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: How are we sure the table is dropped when sweeping?

Maybe the `progress` table just failed to updateand hence it is not null but contains the update from the previous iteration. 
Shouldnt we check `tables.contains(tableToSweep.get())) == false` here? 

I know this was the behavior in the past but now that we are collecting these outcomes and alerting based on these, we should be sure that the outcome recorded is true.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java (hunk 1 above)
Comment (hsaraogi) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Just realised the runtimeexception is caught above. It might be better to catch it here, but its fine in the caller too.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 3]
```diff
@@ -208,26 +231,25 @@ boolean runOnce() {
...
+                return SweepOutcome.TABLE_DROPPED_WHILE_SWEEPING;
             }
-        } catch (RuntimeException e) {
-            log.error("Failed to check whether the table being swept was dropped."
-                    + " Continuing under the assumption that it wasn't...", e);
-            return false;
+        } catch (RuntimeException newE) {
+            log.error("Sweep failed", originalException);
+            log.error("Failed to check whether the table being swept was dropped. Retrying...", newE);
+            return SweepOutcome.ERROR;
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Note that we would RETRYING_WITH_SMALLER_BATCH here too in the past.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 4]
```diff
@@ -110,60 +114,79 @@ public void run() {
...
         }
     }
 
-    // Returns milliseconds to sleep
+    private void updateBatchSize(SweepOutcome outcome) {
+        if (outcome == SweepOutcome.SUCCESS) {
+            batchSizeMultiplier = Math.min(1.0, batchSizeMultiplier * 1.01);
+            return;
+        }
+        if (outcome == SweepOutcome.RETRYING_WITH_SMALLER_BATCH) {
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: What if the outcome is ERROR?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 5]
```diff
@@ -110,60 +114,79 @@ public void run() {
...
+            SweepBatchConfig lastBatchConfig = specificTableSweeper.getAdjustedBatchConfig();
+
+            // Cut batch size in half, always sweep at least one row (we round down).
+            batchSizeMultiplier = Math.max(batchSizeMultiplier / 2, 1.5 / lastBatchConfig.candidateBatchSize());
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: We would change the multiplier after the first time in the past but I think it makes more sense to do it before, just noting that this behavior has changed.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 6]
```diff
@@ -110,60 +114,79 @@ public void run() {
...
     }
 
-    // Returns milliseconds to sleep
+    private void updateBatchSize(SweepOutcome outcome) {
+        if (outcome == SweepOutcome.SUCCESS) {
+            batchSizeMultiplier = Math.min(1.0, batchSizeMultiplier * 1.01);
+            return;
+        }
+        if (outcome == SweepOutcome.RETRYING_WITH_SMALLER_BATCH) {
+            SweepBatchConfig lastBatchConfig = specificTableSweeper.getAdjustedBatchConfig();
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: This is named confusingly, this is `newBatchConfig`?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 7]
```diff
@@ -110,60 +114,79 @@ public void run() {
...
+            batchSizeMultiplier = Math.max(batchSizeMultiplier / 2, 1.5 / lastBatchConfig.candidateBatchSize());
+
+            log.warn("The background sweep job failed unexpectedly with candidate batch size {},"
+                            + " delete batch size {},"
+                            + " and {} cell+timestamp pairs to examine."
+                            + " Attempting to continue with new batchSizeMultiplier {}",
+                    SafeArg.of("candidateBatchSize", lastBatchConfig.candidateBatchSize()),
+                    SafeArg.of("deleteBatchSize", lastBatchConfig.deleteBatchSize()),
+                    SafeArg.of("maxCellTsPairsToExamine", lastBatchConfig.maxCellTsPairsToExamine()),
+                    SafeArg.of("batchSizeMultiplier", batchSizeMultiplier));
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: We already adjusted the batch config, so sweep did not fail with this one as we will run sweep with this config only in the next iteration. The config used in the previous iteration will be `specificTableSweeper.getSweepBatchConfig()`.

Also, logging the batch size multiplier is not great info, lets log the `newBatchConfig` here.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java (hunk 5 above)
Comment (hsaraogi) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Actually we should change it later, otherwise `specificTableSweeper.getSweepBatchConfig()` will not return the previous sweep batch configs for logging.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java (hunk 4 above)
Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: I think with the ERROR outcome something more serious is wrong than just the batch size - so this is unlikely to help.  But to avoid an unnecessary behaviour change I'll also shrink the batch on ERROR.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java (hunk 6 above)
Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: This is kind of funky.  `getSweepBatchConfig` gets the 'base' config values.  `getAdjustedBatchConfig` get the `SweepBatchConfig` but adjusted by applying the `batchSizeMultiplier`.  It doesn't change the underlying config though.

This isn't the `newBatchConfig` though.  This is what we believe the config we ran with for the last batch was.  We haven't yet changed the `batchSizeMultiplier`.  This may be wrong though if the batch config is refreshed in the background.

Basically, we're not passing the actual config we used around, so this line is trying to recreate what was just used.

There's probably a cleaner way of doing this.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java (hunk 5 above)
Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: All of the logging that needs the old config either happens before here or (in the line below) is using the already computed `lastBatchConfig `.

Is there some other behaviour I've accidentally changed here?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java (hunk 7 above)
Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: We haven't changed the batch config here.  We've changed the multiplier but we haven't recalculated the config.  We get the new config if we call `getAdjustedBatchConfig` after changing the multiplier which doesn't happen until the next iteration runs.

This is why mutability sucks, makes it harder to reason about this kind of thing.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java (hunk 3 above)
Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: done

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 8]
```diff
@@ -250,4 +272,28 @@ public synchronized void shutdown() {
...
+        NOT_ENOUGH_DB_NODES_ONLINE, TABLE_DROPPED_WHILE_SWEEPING, ERROR
+    }
+
+    private class SweepOutcomeMetrics {
+        private final MetricsManager metricsManager = new MetricsManager();
+        private SweepOutcome lastOutcome = null;
+
+        SweepOutcomeMetrics() {
+            Arrays.stream(SweepOutcome.values()).forEach(outcome ->
+                    metricsManager.registerMetric(BackgroundSweeperImpl.class, outcome.name(), () -> getCount(outcome))
```
[the remark this answers is not in the document]
Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Yeah, this was a simplification.  We lose some information by only sampling the last result rather than reporting all results.  This is enough for spotting "is sweep disabled for long periods of time" but bad for finer grained detail.  Is it interesting to know that sweep was disabled for 25 seconds?

Things get more complicated if we want to record counts for the outcomes in the period.  Firstly not sure if we can assume there is only 1 collector running.  If more than 1 is running we can't zero the metrics when the lambda is called.

We could use a `Meter` but then we get weird decay functions on the numbers so we see something of the shape of the outcomes but it is difficult to tie back to actual numbers.  Alternatively we can keep records so we can always report "this is the number of each outcome in the last X minutes".

We should chat tomorrow about goal here.  Whatever we do I think I should add a single logging line for "ran iteration and outcome was X" so the fine grained information is clearly visible in the logs.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 9]
```diff
@@ -250,4 +272,28 @@ public synchronized void shutdown() {
...
+            );
+        }
+
+        private Integer getCount(SweepOutcome outcome) {
```
[the remark this answers is not in the document]
Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: done

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java (hunk 2 above)
Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Yeah, I as trying to avoid changing behaviour here but it does look like a bit of leap to say the table must have been dropped here.

I could change the logic to:
```
if(progress.isPresent && !tables.contains(progress.getTable)) {
    return TABLE_DROPPED_WHILE_SWEEPING;
} else {
    return RETRYING_WITH_SMALLER_BATCH;
}
```

At which point I might remove the `RETRYING_WITH_SMALLER_BATCH` outcome and just `ERROR` - we end up retrying with a smaller batch anyway.

Thoughts?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 10]
```diff
@@ -208,26 +231,32 @@ boolean runOnce() {
...
+            log.error("Sweep failed", originalException);
+            log.error("Failed to check whether the table being swept was dropped. Retrying...", newE);
+            return SweepOutcome.ERROR;
+        }
+    }
+
+    private void clearSweepProgress() {
+        Optional<SweepProgress> progress = specificTableSweeper.getTxManager().runTaskReadOnly(
+                specificTableSweeper.getSweepProgressStore()::loadProgress);
+        if (progress.isPresent()) {
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Do we need to load all these values before the clear? Note that #2552 implies we might be loading a lot of data unnecessarily, calling deleteRange even if progress is empty doesnt seem to be harmful, given that we have decided to clear it by this time.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/AdjustableSweepBatchConfigSource.java [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        batchSizeMultiplier = Math.min(1.0, batchSizeMultiplier * 1.01);
+    }
+
+    public void decreaseMultiplier() {
+        SweepBatchConfig lastBatchConfig = getAdjustedSweepConfig();
+
+        // Cut batch size in half, always sweep at least one row (we round down).
+        batchSizeMultiplier = Math.max(batchSizeMultiplier / 2, 1.5 / lastBatchConfig.candidateBatchSize());
+
+        log.warn("The background sweep job failed unexpectedly with candidate batch size {},"
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/AdjustableSweepBatchConfigSource.java]: nit: Seems odd that the AdjustableSweepBatchConfigSource is logs for the `background sweep job`, but wouldnt block on the refactor here.

**Comment by tboam** [comment]
Have addressed PR comments.  Am using a Reservoir to store all the outcomes in the last period of time.  Period is currently hard coded to 1 minute which is the default frequency for reporting and is easy to understand - we could make it configurable in the future if we need to.

We don't compute metrics across all the values as these would be meaningless.

We filter the reservoir to a particular outcome and count how often it happened.  This makes it easy to limit to a fixed window, given sweep is going to take several seconds to run we expect << 100 outcomes per minute so don't think perf is an issue here.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java [hunk 12]
```diff
@@ -250,4 +277,32 @@ public synchronized void shutdown() {
...
+            Arrays.stream(SweepOutcome.values()).forEach(outcome ->
+                    metricsManager.registerMetric(BackgroundSweeperImpl.class, outcome.name(),
+                            () -> getOutcomeCount(outcome))
+            );
+            reservoir = new SlidingTimeWindowReservoir(60L, TimeUnit.SECONDS);
+        }
+
+        private Long getOutcomeCount(SweepOutcome outcome) {
+            return Arrays.stream(reservoir.getSnapshot().getValues())
+                    .filter(l -> l == outcome.ordinal())
```
Comment (hsaraogi) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: This is a bit strange, the metrics now report how many times an outcome was reported in the last minute, so SUCCESS will be 2 for some minute, 3 the next minute, 10 the next minute (if the batches are getting smaller/sweep is getting faster as backup just finished etc.), 

DISABLED will be reported much more frequently and the number could be 100?1000?

This could be useful info as we know how many times each status was reported in the last minute but this info is partially captured by SweepError metrics. Also, this makes setting up alerts a bit tricky as SUCCESS >=1 is success and NOT_ENOUGH_DB_NODES_ONLINE>=1 is we have at least one C* node unhealthy.

I don't directly see the benefit of reporting counts here and what kind of alerts we can have around them, but I maybe missing something. Should we make this 0-1 based on whether the metric was reported in the last minute?

Comment (hsaraogi) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Also, for the disabled case, we can cause some memory overhead. See warning: http://metrics.dropwizard.io/3.1.0/manual/core/#man-core-histograms

Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: I don't think we need to be concerned about the memory overhead here.  We record a single integer each time sweep completes.  If sweep is disabled or the lock is out we pause for 2 minutes between runs - so we're actually only storing a single value.

If sweep succeeds we still sleep for 5 seconds so we're recording 12 metrics per second.  Maybe we'll reduce the sleep and speed up sweep - but we're a long way from recording 1000s of metrics between trimming the backing store.

Comment (tboam) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/BackgroundSweeperImpl.java]: Having the counts allows us to see exactly what sweep has done.  Alarms can then be created for:
- success == 0 for all of the last 24 hours => no sweep run has succeeded, something is very bad
- db_unavailable > 0 for all of the last hour => sweep can't run because a db node is unhealthy and needs attention
- minimum across all hosts of lock_unavailable > 0 for more than 30 minutes => no node could get the sweep lock, implies it hasn't been properly cleared.

Since we can compare to 0, reporting just 0 or 1 doesn't give us anything over reporting the count and just loses information.
